pxtlib/util.ts

```diff
@@ -1804,4 +1804,387 @@ namespace ts.pxtc.BrowserImpl {
+                    return;
+                }
+            }
+            (arr as any)[key] = value;
+        },
+        replace: (arr: any, key: string | number, value: any) => {
+            if (!Array.isArray(arr)) throw new Error("jsonPatch: expected array type");
+            if (typeof key === 'number') {
+                if (key === Math.floor(key)) {
+                    arr.splice(key, 1, value);
+                    return;
+                }
+            }
+            (arr as any)[key] = value;
+        },
+        remove: (arr: any, key: string | number) => {
+            if (!Array.isArray(arr)) throw new Error("jsonPatch: expected array type");
+            if (typeof key === 'number') {
+                if (key === Math.floor(key)) {
+                    arr.splice(key, 1);
+                    return;
+                }
+            }
+            delete (arr as any)[key];
+        },
+    }
+
+    /**
+     * Returns the diff of two objects as a set of change operations, following the
+     * "JSON Patch" format: https://datatracker.ietf.org/doc/html/rfc6902 (with a
+     * small difference to the way paths are encoded).
+     */
+    export function diff(oldObj: any, newObj: any): PatchOperation[] {
+        const diff: {
+            add: AddOperation[]; remove: RemoveOperation[]; replace: ReplaceOperation[];
+        } = {
+            add: [], remove: [], replace: []
+        };
+
+        function _resolveKey(refObj: any, key: string): string | number {
+            if (!Array.isArray(refObj)) return key;
+            const nkey = Number.parseFloat(key);
+            return Number.isNaN(nkey) ? key : nkey;
+        }
+
+        function _diff(oldObj: any, newObj: any, basePath: (string | number)[]) {
+            if (Object.is(oldObj, newObj)) { return; }
+            newObj = newObj || {};
+            oldObj = oldObj || {};
+
+            for (let baseKey of Object.keys(oldObj)) {
+                if (!(baseKey in newObj)) {
+                    const key = _resolveKey(oldObj, baseKey);
+                    // key exists in oldObj but not in newObj -> remove op
+                    diff.remove.push({
+                        op: 'remove',
+                        path: basePath.concat(key)
+                    });
+                }
+            }
+
+            for (const baseKey of Object.keys(newObj)) {
+                const oldVal = oldObj[baseKey];
+                const newVal = newObj[baseKey];
+                const key = _resolveKey(newObj, baseKey);
+                if (!(key in oldObj)) {
+                    if (newObj[key] !== undefined) {
+                        // key exists in newObj but not in oldObj -> add op
+                        if (basePath.length) {
+                            diff.add.push({
+                                op: 'add',
+                                path: basePath,
+                                value: Array.isArray(newObj) ? [] : {}
```

Comment:
should this be newVal?

Reply:
This creates the container newVal will live in. newVal itself is added below.